src/three-components/PMREMGenerator.ts

```diff
@@ -0,0 +1,367 @@
+
+    this.objects.forEach((object) => {
+      this.packingScene.remove(object);
+    });
+    this.cubeLods.forEach((target) => {
+      target.dispose();
+    });
+    this.cubeLods = [];
+  }
+
+  getMipmapShader() {
+    var shaderMaterial = new ShaderMaterial({
+
+      uniforms: {'texelSize': {value: 0.5}, 'envMap': {value: null}},
+
+      vertexShader: `
+varying vec2 vUv;
+varying vec3 vPosition;
+void main() {
+    vUv = uv;
+    vPosition = position;
+    gl_Position = projectionMatrix * modelViewMatrix * vec4( position, 1.0 );
+}
+`,
+
+      fragmentShader: `
+varying vec2 vUv;
+varying vec3 vPosition;
+uniform float texelSize;
+uniform samplerCube envMap;
+int getFace(vec3 direction) {
+    vec3 absDirection = abs(direction);
+    int face = -1;
+    if (absDirection.x > absDirection.z) {
+      if (absDirection.x > absDirection.y)
+        face = direction.x > 0.0 ? 0 : 3;
+      else
+        face = direction.y > 0.0 ? 1 : 4;
+    } else {
+      if (absDirection.z > absDirection.y)
+        face = direction.z > 0.0 ? 2 : 5;
+      else
+        face = direction.y > 0.0 ? 1 : 4;
+    }
+    return face;
+}
+vec3 getDirection(vec2 uv, int face) {
+    uv = 2.0 * uv - 1.0;
+    vec3 direction;
+    if (face == 0) {
+      direction = vec3(1.0, uv.y, -uv.x);
+    } else if (face == 1) {
+      direction = vec3(uv.x, 1.0, -uv.y);
+    } else if (face == 2) {
+      direction = vec3(uv, 1.0);
+    } else if (face == 3) {
+      direction = vec3(-1.0, uv.y, uv.x);
+    } else if (face == 4) {
+      direction = vec3(uv.x, -1.0, uv.y);
+    } else {
+      direction = vec3(-uv.x,uv.y, -1.0);
+    }
+    return direction;
+}
+void main() {
+  int face = getFace(vPosition);
+  vec2 uv = vUv - 0.5 * texelSize;
+  vec3 texelDir = getDirection(uv, face);
+  vec3 color = envMapTexelToLinear(textureCube(envMap, texelDir)).rgb;
+  uv.x += texelSize;
+  texelDir = getDirection(uv, face);
+  color += envMapTexelToLinear(textureCube(envMap, texelDir)).rgb;
+  uv.y += texelSize;
+  texelDir = getDirection(uv, face);
+  color += envMapTexelToLinear(textureCube(envMap, texelDir)).rgb;
+  uv.x -= texelSize;
+  texelDir = getDirection(uv, face);
+  color += envMapTexelToLinear(textureCube(envMap, texelDir)).rgb;
+  gl_FragColor = linearToOutputTexel(vec4(color * 0.25, 1.0));
+}
+`,
+
+      blending: NoBlending
+
+    });
+
+    shaderMaterial.type = 'PMREMGenerator';
+
+    return shaderMaterial;
+  }
+
+  getPackingShader() {
+    var shaderMaterial = new ShaderMaterial({
+
+      uniforms: {
+        'texelSize': {value: 0.5},
+        'envMap': {value: null},
+        'faceIndex': {value: 0},
+      },
+
+      vertexShader: `
+varying vec2 vUv;
+void main() {
+    vUv = uv;
+    gl_Position = projectionMatrix * modelViewMatrix * vec4( position, 1.0 );
+}
+`,
+
+      fragmentShader: `
+varying vec2 vUv;
+uniform float texelSize;
+uniform samplerCube envMap;
+uniform int faceIndex;
+vec3 getDirection(vec2 uv, int face) {
```

Comment:
Let's factor this helper out and use string interpolation to incorporate it into both shaders that use it.